src/Microsoft.Azure.EventHubs/Primitives/ClientEntity.cs

```diff
@@ -55,6 +63,43 @@ public RetryPolicy RetryPolicy
         /// <returns>The asynchronous operation</returns>
         public abstract Task CloseAsync();
 
+        /// <summary>
+        /// Registers a <see cref="EventHubPlugin"/> to be used with this sender.
+        /// </summary>
+        public virtual void RegisterPlugin(EventHubPlugin eventHubPlugin)
+        {
+            if (eventHubPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
+            }
+            if (this.RegisteredPlugins.Any(p => p.Name == eventHubPlugin.Name))
+            {
+                throw new ArgumentException(nameof(eventHubPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(eventHubPlugin)));
+            }
+            this.RegisteredPlugins.Add(eventHubPlugin);
+        }
+
+        /// <summary>
+        /// Unregisters a <see cref="EventHubPlugin"/>.
+        /// </summary>
+        /// <param name="eventHubPlugin">The <see cref="EventHubPlugin.Name"/> of the plugin to be unregistered.</param>
+        public virtual void UnregisterPlugin(string eventHubPlugin)
+        {
+            if (this.RegisteredPlugins == null)
+            {
+                return;
+            }
+            if (string.IsNullOrWhiteSpace(eventHubPlugin))
+            {
+                throw new ArgumentNullException(nameof(eventHubPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(eventHubPlugin)));
+            }
+            if (this.RegisteredPlugins.Any(p => p.Name == eventHubPlugin))
+            {
+                var plugin = this.RegisteredPlugins.First(p => p.Name == eventHubPlugin);
```

Comment:
>First [](start = 52, length = 5)

Remove all matching instead of first here in case RegisteredPlugins has duplicates.